### PR TITLE
osd/scrub: cache frequently used configuration parameters

### PR DESCRIPTION
**Pull request by ronen-fr**
using the md_config_cacher_t, which is a cache object that registers
itself to the config observer and caches the up-to-date configuration
value.

Signed-off-by: Ronen Friedman <rfriedma@redhat.com>

### CODE REVIEW

File: src/osd/scrubber/pg_scrubber.cc [hunk 1]
```diff
@@ -2838,8 +2848,7 @@ void PgScrubber::preemption_data_t::reset()
 
   m_preemptable = false;
   m_preempted = false;
-  m_left = static_cast<int>(
-    m_pg->cct->_conf.get_val<uint64_t>("osd_scrub_max_preemptions"));
+  m_left = static_cast<uint64_t>(osd_scrub_max_preemptions);
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/osd/scrubber/pg_scrubber.cc:2851]: If not, adding

ValueT get() const {...} method to md_config_cacher_t would be preferable to needing to static_cast at each usage site.

[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc:2851]: My understanding is that (with the current code) one must use the static_cast operator (or just a c-cast)
to trigger the operator method - and that the 'T' used must match that of the declaration.
This is indeed error prone (I got the T wrong twice when coding the few lines here).

I will add a 'value()' method to the md_config_cacher (I prefer 'value', which is used in some similar structures, over get() - which we overuse, e.g. for ref-counts)

[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc]: That's a good point. The existing md_... interface requires the cast, and (if I understand correctly)
requires that the 'T' would match the object declaration. This is unusual and error prone.
I have added a () interface (which I  suspect was what was originally intended).

File: src/common/config_cacher.h [hunk 2]
```diff
@@ -53,6 +53,10 @@ class md_config_cacher_t : public md_config_obs_t {
   operator ValueT() const {
     return value_cache.load();
   }
+
+  ValueT operator()() const {
```
Comment (rzarzynski) [inline src/common/config_cacher.h:57]: Do we really need the user conversion operator after introducing `operator()`? I believe backward compatibility isn't an issue as the cacher has just two usages (three with https://github.com/ceph/ceph/pull/60610) across the entire code base.

**Review by rzarzynski — APPROVED** [review]

File: src/common/config_cacher.h (hunk 2 above)
Comment (ronen-fr) [reply src/common/config_cacher.h:57]: Makes sense to remove it. I'll add commits to modify the users and to delete that interface
(like I mentioned above - I suspect it was always meant to be operator()())

Comment (ronen-fr) [reply src/common/config_cacher.h:53]: @rzarzynski - done

**Comment by ronen-fr** [comment]
jenkins retest this please

**Comment by ronen-fr** [comment]
jenkins retest this please

File: src/common/config_cacher.h (hunk 2 above)
Comment (athanatos) [reply src/common/config_cacher.h:53]: I don't find osd_scrub_chunk_min() to be an intuitive interface.  I'd actually prefer to keep the conversion function (```operator ValueT() const```) and add a named ```ValueT to_value() const``` function.  I'm ok with being overruled if you two find this to be clearer.

Comment (ronen-fr) [reply src/common/config_cacher.h:53]: I think it is much clearer than the cast (which has a surprising behavior), and is the most elegant.
We can probably use '*' (which is common enough, e.g. optional or expected), but in our context - just
a wrapper around a value - I like the '()'.

Comment (rzarzynski) [reply src/common/config_cacher.h:53]: I prefer a single method approach.
Whether `operator*` or `operator()` isn't so important. I guess `*my_cached_configurable` would look slightly more familiar but the feeling isn't strong enough to block and request changes.

**Comment by ronen-fr** [comment]
@rzarzynski - I have modified the new api to be operator*, as discussed above. Thanks

**Comment by ronen-fr** [comment]
jenkins retest this please

**Comment by rzarzynski** [comment]
```
ERROR: Error cloning remote repo 'origin'
```

**Comment by rzarzynski** [comment]
jenkins test make check

**Comment by ronen-fr** [comment]
jenkins test make check

**Comment by ronen-fr** [comment]
jenkins test api

**Comment by ronen-fr** [comment]
Merging based on multiple clean Teuthology runs.

**Pull request merged**